<!-- FroboScout launch file -->

<launch>
	<!-- FroboScout parameters --> 
	<param name="diff_steer_wheel_distance" value="0.5"/>
	<param name="ticks_per_meter_left" value="2627" />
	<param name="ticks_per_meter_right" value="2627" />

	<node pkg="rosbag" name="rosbag_record" type="record" args="-a"/>

	<group ns="fmSensors">
		<node pkg="computer_load_monitor" name="load_monitor" type="computer_load_monitor_node" output="screen" >
			<param name="update_rate" value="5" />
		</node>

		<!-- interface to the Hokuyo lidar -->
		<param name="use_rep_117" value="true"/>
		<node name="hokuyo" pkg="hokuyo_node" type="hokuyo_node" output="screen" >
			<param name="port" value="/dev/ttyACM0"/>
			<param name="frame_id" type="string" value="/laser_link"/> 
			<param name="calibrate_time" value="true"/>
			<remap from="scan" to="/fmInformation/laser_msg"/>
			<param name="min_ang" type="double" value="-0.785" />
			<param name="max_ang" type="double" value="0.785" />
		 </node>

		<!-- extract data from messages transmitted by the VectorNav VN-100 IMU -->
		<node pkg="vectornav_vn100" name="imu_node"  type="vectornav_vn100_node" output="screen">
			<param name="vectornav_vn100_sub" value="/fmData/imu_rx"/>
			<param name="imu_pub" value="/fmInformation/imu"/>
			<param name="frame_id" value="imu_link"/>
			<param name="use_enu" value="true"/>
			<param name="covariance_x" value="0.001"/>
			<param name="covariance_y" value="0.001"/>
			<param name="covariance_z" value="0.001"/>
		 </node>

		<!-- extract data from $GPGGA NMEA messages transmitted by the GPS -->
		<node pkg="gps_gpgga" name="gnss_gpgga" type="gps_gpgga_node" output="screen" respawn="true">
		    <param name="subscribe_topic_id" value="/fmData/gnss_rx" />
			<param name="publish_topic_id" value="/fmInformation/gpgga" />
		</node>

		<!-- append Transverse Mercator projection coordinates to NMEA $GPGGA message -->
		<node pkg="gpgga_to_tranmerc" name="gpgga_to_tranmerc" type="gpgga_to_tranmerc_node" output="screen">
			<param name="gpgga_sub" value="/fmInformation/gpgga" />
			<param name="gpgga_tranmerc_pub" value="/fmInformation/gpgga_tranmerc" />
			<param name="transverse_mercator_a" value="6378137.0" />
			<param name="transverse_mercator_f" value="0.0033528106647474805" /> 
			<param name="transverse_mercator_false_easting" value="500000.0" />
			<param name="transverse_mercator_scale_factor" value="0.9996" />
			<param name="transverse_mercator_origin_latitude" value="0.0" />
			<param name="transverse_mercator_central_meridian" value="9.0" />
			<param name="transverse_mercator_false_northing" value="0.0" />
		</node>
	</group>

	<group ns="fmProcessors">
		<node pkg="sdu_surveying_2014" name="log" type="survey_log.py" output="screen">
		</node>

		<!-- calculate odometry based on encoders and IMU yaw axis orientation -->
		<node pkg="differential_odometry" name="differential_odometry" type="differential_odometry_node" output="screen" >
			<param name="enc_left_sub" value="/fmInformation/enc_left" />
			<param name="enc_right_sub" value="/fmInformation/enc_right" />
			<param name="imu_sub" value="/fmInformation/imu" />
			<param name="odom_reset_sub" value="/fmInformation/odom_reset" />
			<param name="odom_pub" value="/fmKnowledge/odom" />
			<!--param name="odom_pub" value="/fmKnowledge/pose" /-->
			<param name="odom_estimate_frame" value="/odom_combined" />
			<param name="vehicle_frame" value="base_footprint" />
			<param name="encoder_output" value="absolute" />
			<param name="max_ticks_per_update" value="200" />
			<param name="yaw_angle_source" value="imu_angular_velocity" />
			<param name="imu_angular_velocity_yaw_axis" value="-x" />
		</node>

		<!-- estimate robot 2d pose -->
		<node pkg="pose_2d_estimator" name="pose_estimator" type="pose_2d_estimator_node.py" output="screen" >
			<param name="odom_sub" value="/fmKnowledge/odom" />
			<param name="imu_sub" value="/fmInformation/imu" />
			<param name="gga_sub" value="/fmInformation/gpgga_tranmerc" />
			<param name="pose_pub" value="/fmKnowledge/pose" />
			<param name="update_rate" value="50" />
			<param name="publish_relative_pose" type="boolean" value="false" />
			<param name="ekf_initial_guess_easting" value="0.0" /> <!-- [m] if unsure set to 0.0 -->
			<param name="ekf_initial_guess_northing" value="0.0" /> <!-- [m] if unsure set to 0.0 -->
			<param name="ekf_initial_guess_yaw" value="0.0" /> <!-- [rad] if unsure set to 0.0 -->
			<param name="odometry_distance_variance" value="0.000001" /> <!-- [m/update] -->
			<param name="odometry_angular_variance" value="0.00035" /> <!-- [rad/update] -->
		</node>

		<!-- node pkg="sdu_surveying_2014" name="survey_log" type="survey_log.py" output="screen" >
			<param name="cmd_vel_sub" value="/fmCommand/cmd_vel" />
			<param name="wptnav_status_sub" value="/fmInformation/wptnav_status" />
			<param name="imu_sub" value="/fmInformation/imu" />
			<param name="gpgga_tranmerc_sub" value="/fmInformation/gpgga_tranmerc" />
			<param name="lidar_sub" value="/fmSensors/laser_msg" />
		</node -->

		<node pkg="frobit_plot" name="froboscout_plot" type="frobit_plot_node.py" output="screen" >
			<param name="pose_sub" value="/fmKnowledge/pose" />
			<param name="gnss_sub" value="/fmInformation/gpgga_tranmerc" />
			<param name="odom_sub" value="/fmKnowledge/odom" />
			<param name="wptnav_status_sub" value="/fmInformation/wptnav_status" />
			<param name="plot_pose_track" value="true" />
			<param name="plot_gnss_track" value="false" />
			<param name="plot_odometry_track" value="false" />
			<param name="plot_pose_yaw" value="false" />
			<param name="map_trackpoint_threshold" value="0.02" />
			<param name="map_max_trackpoints" value="500" />
			<param name="map_minimum_size" value="4.0" />
			<param name="map_easting_offset" value="0.0" />
			<param name="map_northing_offset" value="0.0" />
			<param name="map_update_frequency" value="2" />
			<param name="map_title" value="Frobit track" />
			<param name="map_window_size" value="8.0" />
			<param name="avatar_extension_front" value="0.19" />
			<param name="avatar_extension_rear" value="0.505" />
			<param name="avatar_extension_lateral" value="0.30" />
		</node>
	</group>

	<group ns="fmDecisionMakers">
		<node pkg="sdu_surveying_2014" name="mission" type="wiimote_mission_node.py" output="screen">
			<param name="wiimote_sub" value="/fmLib/joy" />
			<param name="automode_pub" value="/fmDecision/automode" />
    		<param name="deadman_pub" value="/fmCommand/deadman" />
			<param name="cmd_vel_pub" value="/fmCommand/cmd_vel" />
            <param name="max_linear_velocity" value="1.1"/>
            <param name="max_angular_velocity" value="0.5"/>
            <param name="max_linear_acceleration" value="0.2"/>
            <param name="max_angular_acceleration" value="0.3"/>
            <param name="max_linear_deceleration" value="0.8"/>
            <param name="max_angular_deceleration" value="2.0"/>
		</node>

		<!--node pkg="sdu_surveying_2014" name="mission" type="keyboard_mission_node.py" output="screen">
			<param name="keyboard_sub" value="/fmHMI/keyboard" />
			<param name="automode_pub" value="/fmDecision/automode" />
    		<param name="deadman_pub" value="/fmCommand/deadman" />
			<param name="cmd_vel_pub" value="/fmCommand/cmd_vel" />
            <param name="max_linear_velocity" value="0.8"/>
            <param name="max_angular_velocity" value="0.6"/>
            <param name="linear_velocity_step" value="0.1"/>
            <param name="angular_velocity_step" value="0.1"/>
		</node-->
	</group>

	<group ns="fmExecutors">
		<node pkg="sdu_surveying_2014" type="waypoint_navigation_node.py" name="wptnav" output="screen">
			<param name="automode_sub" value="/fmDecision/automode" />
			<param name="pose_sub" value="/fmKnowledge/pose" />
			<param name="joy_sub" value="/fmLib/joy" />
			<param name="cmd_vel_pub" value="/fmCommand/cmd_vel" />
			<param name="implement_pub" value="/fmCommand/implement" />
			<param name="status_pub" value="/fmInformation/wptnav_status" />
			<param name="pid_pub" value="/fmInformation/wptnav_pid" />
			<param name="status_publish_interval" value="1" />
			<param name="pid_publish_interval" value="1" />
			<param name="print_debug_information" value="false" />

            <param name="max_linear_velocity" value="1.2"/>
            <param name="max_angular_velocity" value="0.70"/>

            <param name="drive_kp" value="0.90"/>
            <param name="drive_ki" value="0.20"/>
            <param name="drive_kd" value="0.25"/>
            <param name="drive_feed_forward" value="0.00"/>
            <param name="drive_max_output" value="0.40"/>

            <param name="turn_kp" value="0.6"/>
            <param name="turn_ki" value="0.4"/>
            <param name="turn_kd" value="0.1"/>
            <param name="turn_feed_forward" value="0.0"/>
            <param name="turn_max_output" value="1.0"/>

			<param name="wpt_default_mode" value="MCTE" />
			<param name="wpt_default_tolerance" value="0.10" />
			<param name="wpt_default_drive_velocity" value="0.9" />
			<param name="wpt_default_turn_velocity" value="0.6" />
			<param name="wpt_default_wait_after_arrival" value="5.0" />
			<param name="wpt_default_implement_command" value="0.0" />

			<param name="turn_start_at_heading_err" value="55.0" />
			<param name="turn_stop_at_heading_err" value="2.0" />
			<param name="ramp_drive_velocity_at_distance" value="2.2" />
			<param name="ramp_min_drive_velocity" value="0.15" />
			<param name="ramp_turn_velocity_at_angle" value="25.0" />
			<param name="ramp_min_turn_velocity" value="0.13" />
			<param name="stop_navigating_at_distance" value="0.01" />
		</node>		
	</group>

	<group ns="fmControllers">
	</group>

	<group ns="fmLib">
		<!--node pkg="plot_propulsion_feedback" name="plot_feedback" type="plot_propulsion_feedback_node.py" output="screen" >
 			<param name="propulsion_feedback_sub" value="/fmInformation/wheel_feedback_right" />
 			<param name="update_frequency" value="50" />
			<param name="plot_title" value="Wheel feedback" />
			<param name="maximum_velocity" value="1.2" />
			<param name="maximum_thrust" value="511" />
		</node-->

		<node pkg="int_stamped_to_encoder" name="enc_conversion" type="encoder_converter.py">
			<param name="left_sub" value="/fmInformation/encoder_left" />
			<param name="right_sub" value="/fmInformation/encoder_right" />
			<param name="left_pub" value="/fmInformation/enc_left" />
			<param name="right_pub" value="/fmInformation/enc_right" />
		</node>

		<!-- FroboScout interface node -->
		<node pkg="cetus_froboscout" name="froboscout" type="froboscout_interface_node.py" output="screen" >
    		<param name="deadman_sub" value="/fmCommand/deadman" />
			<param name="cmd_vel_sub" value="/fmCommand/cmd_vel" />
			<param name="enc_left_pub" value="/fmInformation/encoder_left" />
			<param name="enc_right_pub" value="/fmInformation/encoder_right" />
			<param name="wheel_status_left_pub" value="/fmInformation/wheel_status_left" />
			<param name="wheel_status_right_pub" value="/fmInformation/wheel_status_right" />
			<param name="wheel_feedback_left_pub" value="/fmInformation/wheel_feedback_left" />
			<param name="wheel_feedback_right_pub" value="/fmInformation/wheel_feedback_right" />
			<param name="wheel_left_sub" value="/fmData/wheel_left_nmea_in" />	 
			<param name="wheel_left_pub" value="/fmSignal/wheel_left_nmea_out" /> 
			<param name="wheel_right_sub" value="/fmData/wheel_right_nmea_in" />	 
			<param name="wheel_right_pub" value="/fmSignal/wheel_right_nmea_out" /> 
			<param name="publish_wheel_status_rate" value="10" />
			<param name="publish_wheel_feedback_rate" value="50" />
            <param name="max_linear_acceleration" value="1.2"/>
            <param name="max_angular_acceleration" value="2.0"/>
			<param name="wheel_left_kp" value="9" /> 
			<param name="wheel_left_ki" value="15" /> 
			<param name="wheel_left_kd" value="2" /> 
			<param name="wheel_left_max_integral_output" value="50.0" /> 
			<param name="wheel_right_kp" value="9" /> 
			<param name="wheel_right_ki" value="15" /> 
			<param name="wheel_right_kd" value="2" /> 
			<param name="wheel_right_max_integral_output" value="50.0" /> 
			<param name="min_supply_voltage" value="12.0" /> 
		</node>

		<!-- Wiimote interface -->
		<node pkg="wiimote" type="wiimote_node.py" name="wiimote" output="screen" respawn="true">
			<param name="fixed_wiimote_address" value="00:1E:35:03:B8:81"/> 
		</node>		

	 	<!--node pkg="keyboard_interface" type="keyboard_node" name="keyboard">
			<param name="keyboard_pub" value="/fmHMI/keyboard" />
		</node-->

		<!-- nmea protocol handler for the FroboScout left wheel module -->
		<node pkg="serial_nmea" name="left_wheel_nmea" type="nmea_node" output="screen" respawn="true">
			<param name="str_to_msg_sub" value="/fmData/wheel_left_rx" />	 
			<param name="str_to_msg_pub" value="/fmData/wheel_left_nmea_in" />	 
			<param name="msg_to_str_sub" value="/fmSignal/wheel_left_nmea_out" /> 
			<param name="msg_to_str_pub" value="/fmData/wheel_left_tx" />	 
		</node> 

		<!-- nmea protocol handler for the FroboScout right wheel module -->
		<node pkg="serial_nmea" name="right_wheel_nmea" type="nmea_node" output="screen" respawn="true">
			<param name="str_to_msg_sub" value="/fmData/wheel_right_rx" />	 
			<param name="str_to_msg_pub" value="/fmData/wheel_right_nmea_in" />	 
			<param name="msg_to_str_sub" value="/fmSignal/wheel_right_nmea_out" /> 
			<param name="msg_to_str_pub" value="/fmData/wheel_right_tx" />	 
		</node> 

		<!-- serial communication with the FroboScout left wheel module -->
		<node pkg="serial_string" name="left_wheel_ser" type="serial_string_node" output="screen" respawn="true">
			<param name="publisher_topic" value="/fmData/wheel_left_rx" />	 
			<param name="subscriber_topic" value="/fmData/wheel_left_tx" />
			<param name="device" value="/dev/wheel_left" />	
			<param name="baudrate" value="57600" />
			<param name="termination_character" value="10"/>
		</node> 

		<!-- serial communication with the FroboScout right wheel module -->
		<node pkg="serial_string" name="right_wheel_ser" type="serial_string_node" output="screen" respawn="true">
			<param name="publisher_topic" value="/fmData/wheel_right_rx" />	 
			<param name="subscriber_topic" value="/fmData/wheel_right_tx" />
			<param name="device" value="/dev/wheel_right" />	
			<param name="baudrate" value="57600" />
			<param name="termination_character" value="10"/>
		</node> 

		<!-- serial communication with the IMU -->
		<node pkg="serial_string" name="imu_serial" type="serial_string_node" output="screen" respawn="true">
			<param name="publisher_topic" value="/fmData/imu_rx" />	 
			<param name="subscriber_topic" value="/fmData/imu_tx" />
			<param name="device" value="/dev/imu" />	
			<param name="baudrate" value="115200" />
			<param name="termination_character" value="10"/>
		</node> 	

		<!-- serial communication with the RTK-GNSS -->
		<node pkg="serial_string" name="gnss_serial" type="serial_string_node" output="screen" respawn="true">
			<param name="publisher_topic" value="/fmData/gnss_rx" />	 
			<param name="subscriber_topic" value="/fmData/gnss_tx" />
			<param name="device" value="/dev/gps" />	
			<param name="baudrate" value="115200" />
			<param name="termination_character" value="10"/>
		</node>

		<!--node pkg="real_time_analyze" name="rt_node" type="real_time_analyze_node.py" output="screen" >
			<param name="timing_pub" value="/fmInformation/rt_timing"/>
			<param name="device" value="/dev/rt_analyze" />	
        </node--> 	
	</group>
</launch>

